Выполнена задача спринта №20

Jahnta opened the pull request:
Переделана реализация debounce для отправки поискового запроса и нажатия на элементы списка. 
Обновлена реализация отображения прогресса воспроизведения на экране «Аудиоплеер».
Выполнение сетевого запроса для поиска треков реализовано с использованием корутин и Flow на всех уровнях архитектуры исходного кода.

Jahnta committed:
Переделана реализация Debounce для поискового запроса и нажатия на элементы списка. Обновлена реализация отображения прогресса воспроизведения на экране "Аудиоплеер"

Jahnta committed:
Выполнение сетевого запроса для поиска треков реализовано с использованием корутин и Flow на всех уровнях архитектуры исходного кода

i-masloed commented on app/src/main/java/com/practicum/playlistmaker/data/search/impl/SearchRepositoryImpl.kt:
```diff
@@ -24,16 +26,15 @@ class SearchRepositoryImpl(
+        when (response.resultCode) {
             -1 -> {
-                Resource.Error(message = "Проверьте подключение к интернету")
+                emit(Resource.Error(message = "Проверьте подключение к интернету"))
```
все строки должны быть вынесены в ресурсы

i-masloed commented on app/src/main/java/com/practicum/playlistmaker/data/search/network/RetrofitNetworkClient.kt:
```diff
@@ -13,17 +15,22 @@ class RetrofitNetworkClient(
+        return withContext(Dispatchers.IO) {
+            try {
+                val response = iTunesService.searchTrack(dto.queryText)
+                response.apply { resultCode = 200 }
```
Критично: Нелогичное решение с установкой кода ответа внутри приложения. Помимо 200 и 500 кодов в ответе могут прийти любые другие коды.

i-masloed commented on app/src/main/java/com/practicum/playlistmaker/data/player/impl/PlayerRepositoryImpl.kt:
```diff
@@ -18,6 +19,8 @@ class PlayerRepositoryImpl(
     private var playerInfo: PlayerInfo = PlayerInfo(PlayerState.STATE_DEFAULT,"00:00")
     private val observers = mutableListOf<PlayerInfoObserver>()
 
+    private var timerJob: Job? = null
```
Рекомендация: не используется, можно удалить

i-masloed commented on app/src/main/java/com/practicum/playlistmaker/domain/search/SearchInteractor.kt:
```diff
@@ -1,9 +1,10 @@
 
 interface SearchInteractor {
-    fun searchTracks(expression: String, consumer: TracksConsumer)
+    fun searchTracks(expression: String): Flow<Pair<List<Track>?, String?>>
```
Interface TracksConsumer  не используется. Можно удалить

Jahnta committed:
Строки в репозитории вынесены в ресурсы.
Исправлен экран плейера: теперь после воспроизведения трека таймер обновляется и кнопка меняет свое состояние
Прогресс воспроизведения обновляется раз в 300 миллисекунд.

i-masloed commented on app/src/main/java/com/practicum/playlistmaker/utils/ErrorType.kt:
```diff
@@ -0,0 +1,6 @@
+package com.practicum.playlistmaker.utils
+
+enum class ErrorType {
```
Идея верная, но пока не используется
